### PR TITLE
Add "find lowest common ancestor" and tests

### PR DESCRIPTION
**Pull request by ra7kar**
Module to find lowest common ancestor in a binary tree and its test file

### CODE REVIEW

File: data_structures/lowest_common_ancestor_test.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+r_2 = 6
+n1_3 = 6  # test 3
+n2_3 = 13
+r_3 = 10
+n1_4 = 8  # test 4
+n2_4 = 9
+r_4 = 8
+n1_5 = 13  # test 5
+n2_5 = 16
+r_5 = 15
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor_test.py:25]: Just put this in the list below. What is the point of making them individual variables?

File: data_structures/lowest_common_ancestor.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        self.data = data
+        self.left = left
+        self.right = right
+
+    def __str__(self) -> str:
+        return "({})".format(self.data)
+
+
+class BinaryTree:
+    """Binary tress."""
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:17]: There is no point in this comment. It doesn't tell the user anything useful.

File: data_structures/lowest_common_ancestor.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    def __init__(self, py_list=[]):
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:19]: Add documentation for these functions.

File: data_structures/lowest_common_ancestor.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    """Find the lowest common ancestor in the given root of the binary tree
+        with two node values present in the binary tree as inputs (n1, n2)
+
+    Args:
+        root (object): Binary tree root object
+        n1 (int): first node value from the supplied binary tree
+        n2 (int): secont node value from the supplied binary tree
+
+    Returns:
+        (int): intiger valu [description]
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:82]: This is not correct description.

File: data_structures/lowest_common_ancestor.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        return ", ".join(out)
+
+
+def find_ancestor(root, n1, n2):
+    """Find the lowest common ancestor in the given root of the binary tree
+        with two node values present in the binary tree as inputs (n1, n2)
+
+    Args:
+        root (object): Binary tree root object
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:77]: What is object? Everything is an object.

File: data_structures/lowest_common_ancestor.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        return None
+    if root.data == n1 or root.data == n2:
+        return root.data
+
+    left_node = find_ancestor(root.left, n1, n2)
+    right_node = find_ancestor(root.right, n1, n2)
+
+    if left_node is not None and right_node is not None:
+        return root.data
+    if left_node is not None:
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:95]: Use else if.

File: data_structures/lowest_common_ancestor.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    Returns:
+        (int): intiger valu [description]
+    """
+
+    if root is None:
+        return None
+    if root.data == n1 or root.data == n2:
+        return root.data
+
+    left_node = find_ancestor(root.left, n1, n2)
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:90]: left_node is data not the node.

Naming is important. Confusing naming results in confusing code.

File: data_structures/lowest_common_ancestor.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    def __str__(self) -> str:
+        out = []
+        for n in self:
+            out.append(str(n.data))
+
+        return ", ".join(out)
+
+
+def find_ancestor(root, n1, n2):
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:72]: n1 and n2 are not nodes. They are values.

File: data_structures/lowest_common_ancestor.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    """Find the lowest common ancestor in the given root of the binary tree
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:73]: This function can be implemented in tree itself.

File: data_structures/lowest_common_ancestor.py (hunk 8 above)
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:72]: The way this function is written makes no sense to me. I am not sure what it even does.

File: data_structures/lowest_common_ancestor.py (hunk 5 above)
Comment (atrifex) [inline data_structures/lowest_common_ancestor.py:77]: Call it subroot.

File: data_structures/lowest_common_ancestor_test.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+
+@pytest.mark.parametrize(
+    "root, n1, n2, result",
+    [
+        (root, n1_1, n2_1, r_1),
```
Comment (atrifex) [inline data_structures/lowest_common_ancestor_test.py:31]: Don't do root as a param. Do the binary tree and you can extract root in the test function itself.

File: data_structures/lowest_common_ancestor_test.py (hunk 1 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor_test.py:25]: changes implemented

File: data_structures/lowest_common_ancestor.py (hunk 2 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:17]: Added class at the end, indicating it a binary tree class

File: data_structures/lowest_common_ancestor.py (hunk 4 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:82]: description was fixed

File: data_structures/lowest_common_ancestor.py (hunk 5 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:77]: changed it to node of the binary tree

File: data_structures/lowest_common_ancestor.py (hunk 6 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:95]: changed done

File: data_structures/lowest_common_ancestor.py (hunk 7 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:90]: changes left_node to left_node_value and right_node to right_node_value

File: data_structures/lowest_common_ancestor.py (hunk 8 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:72]: changes done

File: data_structures/lowest_common_ancestor.py (hunk 5 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor.py:77]: change implemented

File: data_structures/lowest_common_ancestor_test.py (hunk 10 above)
Comment (ra7kar) [reply data_structures/lowest_common_ancestor_test.py:31]: changes implemented

**Comment by atrifex** [comment]
Why has sudoku been deleted?

**Review by atrifex — APPROVED** [review]

**Pull request merged**